app/models/oauth2/authorization_code_base.rb

```diff
@@ -62,30 +62,32 @@ def refresh_authorization!(&blk)
       return BFailure.new(errors: ["Connection refresh has already failed"])
     end
 
-    refresh_token = fetch_refresh_token
+    T.must(self.with_lock do
```

Comment:
nice! A row-level pessimistic lock to prevent 2 jobs from contacting the provider at the same time. Makes this operation atomic. ACID for the win!